File: 백준/Silver/16401. 과자 나눠주기/과자 나눠주기.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] nums = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int m = nums[0];
        int n = nums[1];
        
        int[] lens = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        Arrays.sort(lens);
        
        // 길이가 x인 막대 과자를 선택
        int left = 1;
        int right = 1_000_000_000;
        int answer = 0;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            // 길이가 x인 막대 과자를 몇개 나눌 수 있는지 확인
            int cnt = 0;
            for (int i = 0; i < n; i ++) {
                int dividableCnt = lens[i] / mid;
                cnt += dividableCnt;
            }
            
            // 모두에게 나눠줄 수 있다면 탐색 범위를 (mid + 1, right) 로 변경
            // 모두에게 나눠줄 수 없다면 탐색 범위를 (left, mid - 1) 로 변경
            if (cnt >= m) {
                answer = Math.max(answer, mid);
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        System.out.println(answer);
    }
}

File: 백준/Silver/16401. 과자 나눠주기/과자 나눠주기.py
def get_available_snack_amount(snack_lengths, l):
    cnt = 0
    for snack_length in snack_lengths:
        max_snack_cnt = snack_length // l # 하나의 과자에서 l만큼의 길이를 가지는 과자의 개수
        cnt += max_snack_cnt
    return cnt

m, n = map(int, input().split())
lens = list(map(int, input().split())) # 오름차순 정렬되어있다

# 과자의 길이(l) 결정
# 길이 l의 과자를 m개 만들 수 있는지 확인
# 만약 부족하다면 과자의 길이를 줄인다.
# 만약 그것보다 많이 만들수있으면 과자의 길이를 늘린다.
# m개 이상의 과자를 만들 수 있고 그 중 가장 큰 l값을 찾는것
# 

left = 1
right = max(lens)
ans = 0

while left <= right:
    mid = (left + right) // 2

    snack_amt = get_available_snack_amount(lens, mid)
    # m개 이상 만들수 있는 경우 과자 길이는 (mid + 1, right)
    if snack_amt >= m:
        ans = max(ans, mid)
        left = mid + 1
    # m개를 만들지 못하는 경우 과자 길이는 (left, mid - 1)
    else:
        right = mid - 1

print(ans)